src/python/pants/task/target_restriction_mixins.py

```diff
@@ -0,0 +1,67 @@
+# coding=utf-8
+# Copyright 2018 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import (absolute_import, division, generators, nested_scopes, print_function,
+                        unicode_literals, with_statement)
+
+from pants.task.goal_options_mixin import GoalOptionsRegistrar
+
+
+class HasTransitiveOptionMixin(object):
+  """A mixin for tasks that have a --transitive option.
+
+  Some tasks must always act on the entire dependency closure. E.g., when compiling, one must
+  compile all of a target's dependencies before compiling that target.
+
+  Other tasks must always act only on the target roots (the targets explicitly specified by the
+  user on the command line). E.g., when finding paths between two user-specified targets.
+
+  Still other tasks may optionally act on either the target roots or the entire closure,
+  as the user prefers in each case. E.g., when invoking a linter. This mixin supports such tasks.
+
+  Note that this mixin doesn't actually register the --transitive option. It assumes that this
+  option was registered on the task (either directly or recursively from its goal).
+  """
+
+  def get_targets(self, predicate=None):
+    if self.get_options().transitive:
+      return self.context.targets(predicate)
+    else:
+      return filter(predicate, self.context.target_roots)
+
+
+class HasSkipOptionMixin(object):
+  """A mixin for tasks that have a --skip option.
+
+  Some tasks may be skipped during certain usages. E.g., you may not want to apply linters
+  while developing.  This mixin supports such tasks.
+
+  Note that this mixin doesn't actually register the --skip option. It assumes that this
+  option was registered on the task (either directly or recursively from its goal).
+  """
+
+  def get_targets(self, predicate=None):
+    if self.get_options().skip:
+      self.context.log.info('Skipping {}.'.format(self.options_scope))
+      return []
+    else:
+      return super(HasSkipOptionMixin, self).get_targets(predicate)
+
+
+# Note order of superclasses - it correctly ensures --skip is checked first.
+class HasSkipAndTransitiveOptionsMixin(HasSkipOptionMixin, HasTransitiveOptionMixin):
```

Comment:
Because of the mixin declaration order, I believe only the latter `get_targets` method will take effect.

Reply:
`HasSkipOptionMixin`s `get_targets()` calls `super`, so I think this is correct? Am I missing something? It seems to work in practice.